presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGeneratorContext.java

```diff
@@ -0,0 +1,378 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot.query;
+
+import com.facebook.presto.pinot.PinotColumnHandle;
+import com.facebook.presto.pinot.PinotConfig;
+import com.facebook.presto.pinot.PinotErrorCode;
+import com.facebook.presto.pinot.PinotException;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.pinot.PushdownUtils.checkSupported;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.String.format;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Encapsulates the components needed to construct a PQL query and provides methods to update the current context with new operations.
+ */
+public class PinotQueryGeneratorContext
+{
+    public static final String TIME_BOUNDARY_FILTER_TEMPLATE = "__TIME_BOUNDARY_FILTER_TEMPLATE__";
+    public static final String TABLE_NAME_SUFFIX_TEMPLATE = "__TABLE_NAME_SUFFIX_TEMPLATE__";
+    // Fields defining the query
+    // order map that maps the column definition in terms of input relation column(s)
+    private final LinkedHashMap<VariableReferenceExpression, Selection> selections;
```

Comment:
Use interfaces than implementation

Reply:
I really care about this being order sensitive and want to let the reader be completely aware of that. We have caused bugs before with loosing this information and would very much like to let this remain.

We generate the pinot PQL using this and thus order sensitivity is a first order property that I would like to uplevel to the field type.